common/types/string_test.go

```diff
@@ -253,3 +253,190 @@ func TestStringSize(t *testing.T) {
+			format:         "%%%s%%",
+			formatArgs:     []ref.Val{String("text")},
+			expectedOutput: "%text%",
+		},
+		{
+			name:           "substitution with one escaped percent sign on the right",
+			format:         "%s%%",
+			formatArgs:     []ref.Val{String("percent on the right")},
+			expectedOutput: "percent on the right%",
+		},
+		{
+			name:           "substitution with one escaped percent sign on the left",
+			format:         "%%%s",
+			formatArgs:     []ref.Val{String("percent on the left")},
+			expectedOutput: "%percent on the left",
+		},
+		{
+			name:   "multiple substitutions",
+			format: "%s %s %s, %s %s %s, %s %s %s, %s %s %s",
+			formatArgs: []ref.Val{
+				Int(1), Int(2), Int(3),
+				String("A"), String("B"), String("C"),
+				Int(4), Int(5), Int(6),
+				String("D"), String("E"), String("F"),
+			},
+			expectedOutput: "1 2 3, A B C, 4 5 6, D E F",
+		},
+		{
+			name:   "percent sign escape sequence support",
+			format: "\u0025\u0025escaped \u0025s\u0025\u0025",
+			formatArgs: []ref.Val{
+				String("percent"),
+			},
+			expectedOutput: "%escaped percent%",
+		},
+		{
+			name:   "fixed point formatting clause",
+			format: "%.3f",
+			formatArgs: []ref.Val{
+				Double(1.2345),
+			},
+			expectedOutput: "1.234",
+		},
+		{
+			name:   "binary formatting clause",
+			format: "this is 5 in binary: %b",
+			formatArgs: []ref.Val{
+				Int(5),
+			},
+			expectedOutput: "this is 5 in binary: 101",
+		},
+		{
+			name:   "octal formatting clause",
+			format: "%o",
+			formatArgs: []ref.Val{
+				Int(11),
+			},
+			expectedOutput: "13",
+		},
+		{
+			name:   "lowercase hexadecimal formatting clause",
+			format: "%h is 20 in hexadecimal",
+			formatArgs: []ref.Val{
+				Int(30),
+			},
+			expectedOutput: "1e is 20 in hexadecimal",
+		},
+		{
+			name:   "uppercase hexadecimal formatting clause",
+			format: "%H is 20 in hexadecimal",
+			formatArgs: []ref.Val{
+				Int(30),
+			},
+			expectedOutput: "1E is 20 in hexadecimal",
+		},
+		{
+			name:   "lowercase scientific notation formatting clause",
+			format: "%.6e",
+			formatArgs: []ref.Val{
+				Double(1052.032911275),
+			},
+			expectedOutput: "1.052033\u202f×\u202f10⁰³",
+		},
+		{
+			name:   "uppercase scientific notation formatting clause",
+			format: "%.6E",
+			formatArgs: []ref.Val{
+				Double(1052.032911275),
+			},
+			expectedOutput: "1.052033\u202f×\u202f10⁰³",
+		},
+		{
+			name:           "locale support",
+			format:         "%.3f",
+			formatArgs:     []ref.Val{Double(3.14)},
+			locale:         "fr_FR",
+			expectedOutput: "3,140",
+		},
+		{
+			name:   "unrecognized formatting clause",
+			format: "%a",
+			formatArgs: []ref.Val{
+				Int(1),
+			},
+			err: "could not parse formatting clause: unrecognized formatting clause \"a\"",
+		},
+		{
+			name:       "out of bounds arg index",
+			format:     "%s %s %s",
+			formatArgs: []ref.Val{Int(0), Int(1)},
+			err:        "index 2 out of range",
```

Comment:
I was expecting a lot more err cases. Illegal arg types, too many args, lots of malformed format clauses.

Reply:
Right - I'll add a full test suite tonight. I think most of the happy path cases are covered decently enough; most of the coverage is missing in the error cases.

Reply:
@DangerOnTheRanger before you get too much farther, try to move this functionally into the `ext/strings.go` file since the intent is that this functionally is purely optional, even if it's well defined. If you do need to make some changes to the string type, that's okay, but they should hopefully be minimal.

Reply:
Whoops, you're right. I'll also take care of that tonight as well.

Reply:
This also makes it clear where to put the documentation as `ext/strings.go` has clear documentation conventions.

Reply:
(marking as resolved, tracking test suite in https://github.com/google/cel-go/pull/617#discussion_r1095261962 instead)